Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+    default: Py_UNREACHABLE();
+    }
+}
+
+int
+_PyParkingLot_Park(const void *addr, const void *expected, size_t size,
+                   _PyTime_t timeout_ns, void *arg, int detach)
+{
+    struct wait_entry wait;
+    wait.unpark.arg = arg;
+    wait.addr = (uintptr_t)addr;
+
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (!validate_addr(addr, expected, size)) {
+        _PyRawMutex_Unlock(&bucket->mutex);
+        return Py_PARK_AGAIN;
+    }
+    wait.sema = _PySemaphore_Alloc();
+    enqueue(bucket, addr, &wait);
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+    int res = _PySemaphore_Wait(wait.sema, timeout_ns, detach);
+    if (res == Py_PARK_OK) {
+        goto done;
+    }
+
+    // timeout or interrupt
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (wait.node.next == NULL) {
+        _PyRawMutex_Unlock(&bucket->mutex);
+        // We've been removed the waiter queue. Wait until we process the
+        // wakeup signal.
+        do {
+            res = _PySemaphore_Wait(wait.sema, -1, detach);
+        } while (res != Py_PARK_OK);
+        goto done;
+    }
+    else {
+        llist_remove(&wait.node);
+        --bucket->num_waiters;
+    }
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+done:
+    _PySemaphore_Free(wait.sema);
+    return res;
+
+}
+
+struct _PyUnpark *
+_PyParkingLot_BeginUnpark(const void *addr)
+{
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+    _PyRawMutex_Lock(&bucket->mutex);
+
+    struct wait_entry *waiter = dequeue(bucket, addr);
+    if (!waiter) {
+        return NULL;
+    }
+
+    waiter->unpark.has_more_waiters = (bucket->num_waiters > 0);
+    return &waiter->unpark;
+}
+
+#define container_of(ptr, type, member) \
+    ((type *)((char *)(ptr) - offsetof(type, member)))
+
+void
+_PyParkingLot_FinishUnpark(const void *addr, struct _PyUnpark *unpark)
+{
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+    if (unpark) {
+        struct wait_entry *waiter;
+        waiter = container_of(unpark, struct wait_entry, unpark);
+
+        _PySemaphore_Wakeup(waiter->sema);
+    }
+}
+
+void
+_PyParkingLot_UnparkAll(const void *addr)
+{
+    struct llist_node head = LLIST_INIT(head);
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+
+    _PyRawMutex_Lock(&bucket->mutex);
+    dequeue_all(bucket, addr, &head);
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+    struct llist_node *node;
+    llist_for_each_safe(node, &head) {
+        struct wait_entry *waiter = llist_data(node, struct wait_entry, node);
+        llist_remove(node);
+        _PySemaphore_Wakeup(waiter->sema);
+    }
+}
+
+void
+_PyParkingLot_AfterFork(void)
+{
+    // After a fork only one thread remains. That thread cannot be blocked
+    // so all entries in the parking lot are for dead threads.
+    memset(buckets, 0, sizeof(buckets));
```

Comment:
Is there any advantage to this over letting the compiler do it?

```suggestion
    buckets = (Bucket[NUM_BUCKETS]){{{0}}};
```

Reply:
C doesn't let you assign to arrays (and buckets is an array of structs).

It would work if `buckets` were a struct containing an array, but the code generation is [much worse](https://gcc.godbolt.org/z/74oY4T3nz) on MSVC compared to `memset`, even with optimizations enabled. (It uses a *lot* of stack).